lib/protocol/protocol.js

```diff
@@ -226,41 +235,84 @@ function _decodeFetchResponse (resp, cb, maxTickMessages, version) {
     .word32bs('topicNum')
     .loop(decodeTopics(decodePartitions));
 
+  // Parsing is all sync, but emitting events is async due to compression
+  // At this point, the enqueue chain should be populated in order, and at
+  // the end, the 'topics' map above should be fully populated.
+  // topics is not ready for use until all events are processed, as an async
+  // decompression may add more offsets to the topic map.
+  async.series(events, (err) => {
```

Comment:
Could we process messages in parallel instead of series? Looks like only the results (the decoded messages) need to be emitted in order.

Reply:
No, that would require 2 queues of processing, and complicate the logic greatly, and not really save any meaningful amount of time.

Reply:
It would require a queue per topic/partition combination, and would require waiting until every queue is done processing.

And for most cases, the processing is synchronous anyways. We only go async for compression and hitting max-messages-per-tick.